vnext/Shared/Modules/BlobModule.cpp

```diff
@@ -0,0 +1,218 @@
+using std::mutex;
+using std::string;
+using std::vector;
+using std::weak_ptr;
+using winrt::Windows::Foundation::GuidHelper;
+using winrt::Windows::Security::Cryptography::CryptographicBuffer;
+
+namespace {
+constexpr char moduleName[] = "BlobModule";
+constexpr char blobURIScheme[] = "blob";
+
+weak_ptr<Microsoft::React::IWebSocketModuleContentHandler> s_contentHandler;
+} // namespace
+
+namespace Microsoft::React {
+
+#pragma region BlobModule
+
+BlobModule::BlobModule() noexcept : m_sharedState{std::make_shared<SharedState>()} {
+  m_contentHandler = std::static_pointer_cast<BlobWebSocketModuleContentHandler>(s_contentHandler.lock());
+  if (!m_contentHandler) {
+    m_contentHandler = std::make_shared<BlobWebSocketModuleContentHandler>();
+    s_contentHandler = m_contentHandler;
+  }
+
+  m_sharedState->Module = this;
+}
+
+BlobModule::~BlobModule() noexcept /*override*/ {
+  m_sharedState->Module = nullptr;
+}
+
+#pragma region CxxModule
+
+string BlobModule::getName() {
+  return moduleName;
+}
+
+std::map<string, dynamic> BlobModule::getConstants() {
+  return {{"BLOB_URI_SCHEME", blobURIScheme}, {"BLOB_URI_HOST", {}}};
+}
+
+std::vector<module::CxxModule::Method> BlobModule::getMethods() {
+  return {
+      {"addNetworkingHandler",
+       [](dynamic args) {
+         // TODO: Implement #6081
+       }},
+
+      {"addWebSocketHandler",
+       [contentHandler = m_contentHandler](dynamic args) {
+         auto id = jsArgAsInt(args, 0);
+
+         contentHandler->Register(id);
+       }},
+
+      {"removeWebSocketHandler",
+       [contentHandler = m_contentHandler](dynamic args) {
+         auto id = jsArgAsInt(args, 0);
+
+         contentHandler->Unregister(id);
+       }},
+
+      {"sendOverSocket",
+       // As of React Native 0.67, instance is set AFTER CxxModule::getMethods() is invoked.
+       // Directly use getInstance() once
+       // https://github.com/facebook/react-native/commit/1d45b20b6c6ba66df0485cdb9be36463d96cf182 becomes available.
+       [contentHandler = m_contentHandler, weakState = weak_ptr<SharedState>(m_sharedState)](dynamic args) {
+         auto blob = jsArgAsObject(args, 0);
+         auto blobId = blob["blobId"].getString();
+         auto offset = blob["offset"].getInt();
+         auto size = blob["size"].getInt();
+         auto socketID = jsArgAsInt(args, 1);
+
+         auto data = contentHandler->ResolveMessage(std::move(blobId), offset, size);
+
+         if (auto state = weakState.lock()) {
+           auto buffer = CryptographicBuffer::CreateFromByteArray(data);
+           auto winrtString = CryptographicBuffer::EncodeToBase64String(std::move(buffer));
+           auto base64String = Common::Unicode::Utf16ToUtf8(std::move(winrtString));
+
+           auto sendArgs = dynamic::array(std::move(base64String), socketID);
+           if (auto instance = state->Module->getInstance().lock()) {
+             instance->callJSFunction("WebSocketModule", "sendBinary", std::move(sendArgs));
```

Comment:
Before merging:

Allow invoking this functionality directly on the native thread to avoid an unnecessary round trip.

Reply:
Update.

Will communicate with the `WebSocketModule` via interface `IWebSocketModuleProxy`.